packages/@lwc/template-compiler/src/parser/index.ts

```diff
@@ -882,7 +882,11 @@ export default function parse(source: string, state: State): TemplateParseResult
     }
 
     function validateChildren(element: IRElement) {
-        if (element.lwc?.dom && element.children.length > 0) {
+        if (!element.lwc?.dom) return;
+        const children = getPreserveComments(element)
+            ? element.children
+            : element.children.filter((child) => child.type !== 'comment');
+        if (children.length > 0) {
```

Comment:
As this method is only about the `lwc:dom` directive and since we might add more validation steps later and I would recommend removing the early exit.

```suggestion
        const effectiveChildren = getPreserveComments(element)
            ? element.children
            : element.children.filter((child) => child.type !== 'comment');
        if (element.lwc?.dom && effectiveChildren.length > 0) {
```